refact(ngOptions): major refactoring...

petebacondarwin opened the pull request:
**Major reworking of select and ngOptions**:
- The `SelectController` is now used as an abstraction for the `select` and `ngOptions` directives
  to override to get their desired behaviour
- The `select` directive is completely oblivious to the ngOptions directive now - the `ngOptions`
  directive could be deleted without having to make any changes to the `select` directive.
- Select related directives (single/multiple/ngOptions) can provide specific versions of
  `SelectController.writeValue` and `SelectController.readValue`, which are responsible for getting
  the `$viewValue` in or out of the actual `<select>` element and its `<option>` children.

BREAKING CHANGE:

When using `ngOptions`: the directive applies a surrogate key as the value of the `<option>` element.
This commit changes the actual string used as the surrogate key. We now store a string that is computed
by calling `hashKey` on the item in the options collection; previously it was the index or key of the
item in the collection.

Before you might have seen:

```
<select ng-model="x" ng-option="i in items">
  <option value="1">a</option>
  <option value="2">b</option>
  <option value="3">c</option>
  <option value="4">d</option>
</select>
```

Now it will be something like:

```
<select ng-model="x" ng-option="i in items">
  <option value="string:a">a</option>
  <option value="string:b">b</option>
  <option value="string:c">c</option>
  <option value="string:d">d</option>
</select>
```

If your application code relied on this value, which it shouldn't, then you will need to modify your
application to accommodate this. You may find that you can use the `track by` feaure of `ngOptions`
as this provides the ability to specify the key that is stored.

BREAKING CHANGE:

When iterating over an object's properties using the `(key, value) in obj` syntax the order of the elements used to be sorted alphabetically. This was an artificial attempt to create a deterministic ordering since browsers don't guarantee the order. But in practice this is not what people want and so this change iterates over properties in the order they are returned by `Object.keys(obj)`, which is almost always the order in which the properties were defined.

Closes #8019
Closes #9714
Closes #10531

lgalfaso commented:
would it be possible to split this into two commits, one that move things around and another that make the modifications? It would make it a lot easier to review the patch. One PR is ok and so is the order of the commits

petebacondarwin commented:
Not really. The original file (select.js) is so interwoven it was not really feasible to extract the ngOptions bits without doing a lot of the refactoring. I'll see what I can do...

petebacondarwin commented:
I guess I could combine these two files into one and then split them in a subsequent commit?

lgalfaso commented:
the issue with the current patch is that is very hard to know what actually changed and what was only moved. Whatever makes this easy, works for me

petebacondarwin commented:
Well pretty much everything that was related to ngOptions changed :-)

petebacondarwin committed:
fix(ngOptions): ensure that the correct option is selected when options are loaded async

**Major reworking of select and ngOptions**:

* The `SelectController` is now used as an abstraction for the `select` and `ngOptions` directives
to override to get their desired behaviour
* The `select` directive is completely oblivious to the ngOptions directive now - the `ngOptions`
directive could be deleted without having to make any changes to the `select` directive.
* Select related directives (single/multiple/ngOptions) can provide specific versions of
`SelectController.writeValue` and `SelectController.readValue`, which are responsible for getting
the `$viewValue` in or out of the actual `<select>` element and its `<option>` children.

BREAKING CHANGE:

When using `ngOptions`: the directive applies a surrogate key as the value of the `<option>` element.
This commit changes the actual string used as the surrogate key. We now store a string that is computed
by calling `hashKey` on the item in the options collection; previously it was the index or key of the
item in the collection.

(This is in keeping with the way that the unknown option value is represented in the select directive.)

Before you might have seen:

```
<select ng-model="x" ng-option="i in items">
  <option value="1">a</option>
  <option value="2">b</option>
  <option value="3">c</option>
  <option value="4">d</option>
</select>
```

Now it will be something like:

```
<select ng-model="x" ng-option="i in items">
  <option value="string:a">a</option>
  <option value="string:b">b</option>
  <option value="string:c">c</option>
  <option value="string:d">d</option>
</select>
```

If your application code relied on this value, which it shouldn't, then you will need to modify your
application to accommodate this. You may find that you can use the `track by` feaure of `ngOptions`
as this provides the ability to specify the key that is stored.

BREAKING CHANGE:

When iterating over an object's properties using the `(key, value) in obj` syntax
the order of the elements used to be sorted alphabetically. This was an artificial
attempt to create a deterministic ordering since browsers don't guarantee the order.
But in practice this is not what people want and so this change iterates over properties
in the order they are returned by Object.keys(obj), which is almost always the order
in which the properties were defined.

Closes #8019
Closes #9714
Closes #10639

petebacondarwin commented:
@lgalfaso  - How's that?

lgalfaso commented:
what will happen if the same option value shows up several times? This can happen when the two options are in different option groups

petebacondarwin commented:
>what will happen if the same option value shows up several times? This can happen when the two options are in different option groups

Good question, I have to check the previous implementation but I think it didn't support multiple options with the same value either,

jayoungers commented:
Even if the previous implementation did support it, is that really a valid requirement?  How would you even select the correct item in the list the next time they view that page?  you'd have to 'track by' group and value.

petebacondarwin commented:
@jayoungers - thanks for commenting. I agree that this requirement would be tricky. You could certainly create a `track by` expression that was unique for a given group/value combination.

lgalfaso commented:
I was thinking when there is no `value`, and the text is used

petebacondarwin commented:
@lgalfaso - so you are referring to the standard `select` and `option` directive combination rather than `ngOptions`? I don't know if this was ever supported either since the `addOption` method always took the value of the option as its parameter, which means that you would only get one option in the `SelectController`'s map

lgalfaso commented:
@petebacondarwin ok, makes sense

petebacondarwin commented:
>will using a hashmap impact performance with many options? Generally, extraordinary numbers of options smell of bad ui, but it's possible, see e.g. #8019 - 2500 options is definitely too much, but we should test if there's a visible regression for first render time with say, 200 options.

I need to set up some performance tests for this refactoring to check the impact of these changes. One of the main benefits this change provides is that the option list is no longer touched at all when the view or model are changing - only if the data that defines the options changes.

>In the breaking changes you say that the application shouldn't rely on the value of the option. However if you use ngOptions to render, but submit the form to the server via action="", then your backend will probably rely on the correct select value. I guess the number of people who do this is very low, though.

In `ngOptions` the value of each option is supposed to be a surrogate key - i.e. it should not be relied upon. If people do rely on this then they could recreate it by using a `track by` expression, since the value of each option is the result of evaluating `track by` for each option item.

>now that ngOptions and select are two different directives, could this cause problems of interoperability with other directives? For example, if someone attached a directive to a select that is supposed to run after the select is completely initialized, could this break?

There was actually always an `ngOptions` directive but all it did was terminate the compilation to prevent `<option>` elements from compiling and trying to interact with the `SelectController`. A main feature of this refactoring is that the logic for setting up `ngOption` based selects has moved from the `select` directive to the `ngOptions` directive.

petebacondarwin commented:
The test case I am using looks like this:

```js
<!DOCTYPE html>
<html>

  <head>
    <script src="path/to/angular.js"></script>
    <script type="text/javascript">
      angular.module('app', [])

      .controller('C', function($scope) {
        $scope.items = [];
        $scope.count = 10000;

        $scope.changeOptions = function() {
          $scope.items = [];
          for(var i=0; i<$scope.count; ++i) {
            $scope.items.push({
              id: i,
              label: 'item-' + i,
              group: 'group-' + i % 100
            });
          }
        };
      });
    </script>
  </head>

  <body ng-app="app" ng-controller="C">
    <h1>refactored</h1>
    <input ng-model="count" type="text"><button ng-click="changeOptions()">Change Options</button>
    <br/>
    <select ng-model="x" ng-options="a as a.label group by a.group for a in items track by a.id"></select>
    {{ x }}
  </body>

</html>
```

lgalfaso commented on src/ng/directive/select.js:
```diff
@@ -745,31 +249,30 @@ var optionDirective = ['$interpolate', function($interpolate) {
+              if (oldVal !== newVal) {
+                selectCtrl.removeOption(oldVal);
+              }
+              selectCtrl.addOption(newVal, element);
```
two lines up, we are removing an option (and calling $render) and then adding an option (and calling $render again)

Narretz replied:
Does It make sense to move the call to `$render`out of  `add/removeOption`? With many repeated options you have quite a few calls to $render. Imho, it's also a good idea if the add/remove functions don't do any additional work.

petebacondarwin replied:
$render is now pretty cheap but I think you are probably right here.

petebacondarwin replied:
I fixed this too.

Narretz commented:
@petebacondarwin Do you plan on benchmarking the render time with benchpress? Visually I don't see a huge difference in render time with your template.

gkalpak commented on src/ng/directive/ngOptions.js:
```diff
@@ -0,0 +1,593 @@
+        if (emptyOption_ || unknownOption_) {
+          while (current &&
+                current === emptyOption_ ||
+                current === unknownOption_) {
```
Is this really supposed to mean `(current && current === emptyOption) || current === unknownOption` ?

petebacondarwin replied:
No :-)

gkalpak commented on src/ng/directive/select.js:
```diff
@@ -1,6 +1,113 @@
+  self.removeOption = function(value) {
+    var count = optionsMap.get(value);
+    if (count) {
+      if (count == 1) {
```
Since `optionsMap` only seem to store integer values, shouldn't we use `===` instead ?
Do I miss something ?

petebacondarwin replied:
Yes :-)

gkalpak commented on src/ng/directive/select.js:
```diff
@@ -344,400 +202,46 @@ var selectDirective = ['$compile', '$parse', function($compile,   $parse) {
   return {
     restrict: 'E',
     priority: 100,
+    require: '?^select',
```
Minor, but using `^^` (instead of `^`) would save one level of checking.

petebacondarwin replied:
I forgot about that!

petebacondarwin replied:
Actually Igor originally put in an even more aggressive optimization in the original, which I guess we should use.

gkalpak commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -0,0 +1,1731 @@
+    });
+
+
+    it('should bind to scope value through experession', function() {
```
Type: _experession_

petebacondarwin replied:
+1

gkalpak commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -0,0 +1,1731 @@
+    });
+
+
+    it('It should use the "value" variable to represent items in the array as well as for the ' +
```
Redundant: `It `

gkalpak commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -0,0 +1,1731 @@
+        }
+
+        if (!equals(this.actual.val(), value)) {
+          errors.push('Expected select value "' + this.actual.val() + '" to equal "' + hashKey(value) + '"');
```
Wouldn't this result in re-hashing the value ? Is this intended ?

petebacondarwin replied:
Another good spot! Originally I was hashing inside the if statement but not now.

petebacondarwin commented:
Awesome bit of review @gkalpak - thank you. I will address those issues over the weekend.

petebacondarwin committed:
refact(ngOptions): move into its own file

Since select is not aware of ngOptions, it makes sense to move it into its
own file for more easy maintenance.

petebacondarwin committed:
test(ngOptions): should place non-grouped items in the list where they appear

Closes #10531

petebacondarwin committed:
test(ngOptions): should not insert a blank option if one of the options maps to null

Closes #7605

petebacondarwin committed:
style(ngOptionsSpec): add extra newline for better formatting

petebacondarwin committed:
fix(ngOptions): update model if selected option is removed

Closes #7736

petebacondarwin committed:
refact(ngOptions): specialize readValue and writeValue based on multiple attribute

petebacondarwin committed:
fix(ngOptions): prevent infinite digest if track by expression is stable

Closes #9464

petebacondarwin committed:
fix(ngOptions): support one-time binding on the option values

Utilize the $watchDelegate on the watcher used to detect changes to the labels.

Closes #10687
Closes #10694

lgalfaso commented:
LGTM, @petebacondarwin can you land this?

petebacondarwin commented:
Landing now.
